src/ledger/LedgerTxn.cpp

```diff
@@ -1225,6 +1496,192 @@ LedgerTxn::Impl::maybeUpdateLastModifiedThenInvokeThenSeal(
+{
+    // recordEntry has the strong exception safety guarantee because
+    // - std::unordered_map<...>::erase has the strong exception safety
+    //   guarantee
+    // - std::unordered_map<...>::operator[] has the strong exception safety
+    //   guarantee
+    // - std::shared_ptr<...>::operator= does not throw
+    auto recordEntry = [&]() {
+        if (eraseIfNull)
+        {
+            mEntry.erase(key);
+        }
+        else
+        {
+            mEntry[key] = lePtr;
+        }
+    };
+
+    // If the key does not correspond to an offer, we do not need to manage the
+    // order book. Record the update in mEntry and return.
+    if (key.type() != OFFER)
+    {
+        recordEntry();
+        return;
+    }
+
+    auto mobIterPrev = mMultiOrderBook.end();
+    OrderBook::iterator obIterPrev;
+    auto iter = mEntry.find(key);
+    if (iter != mEntry.end() && iter->second)
+    {
+        std::tie(mobIterPrev, obIterPrev) = findInOrderBook(*iter->second);
+    }
+
+    // We only insert the new offer into the order book if it exists and is not
+    // active. Otherwise, we just record the update in mEntry and return.
+    if (lePtr && !effectiveActive)
+    {
+        auto const& oe = lePtr->data.offer();
+        AssetPair assetPair{oe.buying, oe.selling};
+
+        auto mobIterNew = mMultiOrderBook.find(assetPair);
+        if (mobIterNew != mMultiOrderBook.end())
+        {
+            auto& obNew = mobIterNew->second;
+            auto res = obNew.insert({{oe.price, oe.offerID}, key});
+            try
+            {
+                recordEntry();
+            }
+            catch (...)
+            {
+                obNew.erase(res);
+                throw;
+            }
+        }
+        else
+        {
+            auto res = mMultiOrderBook.emplace(
+                assetPair, OrderBook{{{oe.price, oe.offerID}, key}});
+            try
+            {
+                recordEntry();
+            }
+            catch (...)
+            {
+                mMultiOrderBook.erase(res.first);
+                throw;
+            }
+        }
+    }
+    else
+    {
+        recordEntry();
+    }
+
+    // This never throws
+    if (mobIterPrev != mMultiOrderBook.end() &&
+        obIterPrev != mobIterPrev->second.end())
+    {
+        mobIterPrev->second.erase(obIterPrev);
+    }
+}
+
+void
+LedgerTxn::Impl::updateWorstBestOffer(
+    AssetPair const& assets, std::shared_ptr<OfferDescriptor const> worstBest)
+{
+    // Update mWorstBestOffer if
+    // - assets is currently not in mWorstBestOffer
+    // - entry is nullptr, because if any offers exist they would be returned
+    //   before nullptr
+    // - assets is in mWorstBestOffer, entry is not nullptr, and *entry is
+    //   worse than mWorstBestOffer[assets] because then mWorstBestOffer[assets]
+    //   would be returned before *entry
+    auto iter = mWorstBestOffer.find(assets);
+    if (iter == mWorstBestOffer.end() || !worstBest ||
```

Comment:
per definition of "worse than", should the condition be:
```
if (iter == mWorstBestOffer.end() || !entry ||
        !iter->second || isBetterOffer(*iter->second, *entry))
```

Reply:
I think current code doesn't override `null` because of this, which I think is a (performance only) bug